### PR TITLE
oauth access token mapping resource

### PR DESCRIPTION
**Pull request by wesleymccollam**
- oauth access token mapping resource
- `location` marked as deprecated

### CODE REVIEW

File: internal/acctest/config/oauth/accesstokenmapping/oauth_access_token_mapping_test.go [hunk 1]
```diff
@@ -0,0 +1,352 @@
+package acctest_test
```
Comment (henryrecker-pingidentity) [inline internal/acctest/config/oauth/accesstokenmapping/oauth_access_token_mapping_test.go:1]: This package should be accesstokenmapping_test

File: internal/acctest/config/oauth/accesstokenmapping/oauth_access_token_mapping_test.go [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+			},
+			{
+				// Test importing the resource
+				Config:            testAccOauthAccessTokenMapping(resourceName, updatedResourceModel),
+				ResourceName:      "pingfederate_oauth_access_token_mapping." + resourceName,
+				ImportStateId:     oauthAccessTokenMappingId,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+			{
```
Comment (henryrecker-pingidentity) [inline internal/acctest/config/oauth/accesstokenmapping/oauth_access_token_mapping_test.go:86]: Should we have a delete drift check in here?

File: internal/resource/common/attributesources/attribute_sources_schema.go [hunk 3]
```diff
@@ -17,12 +17,12 @@ import (
...
 )
 
-func CommonAttributeSourceSchema() map[string]schema.Attribute {
+func CommonAttributeSourceSchema(optionalAndComputedNestedAttributeContractFulfillment bool) map[string]schema.Attribute {
 	commonAttributeSourceSchema := map[string]schema.Attribute{}
 	commonAttributeSourceSchema["data_store_ref"] = schema.SingleNestedAttribute{
 		Required:    true,
 		Description: "Reference to the associated data store.",
-		Attributes:  resourcelink.ToSchema(),
+		Attributes:  resourcelink.ToSchemaLocationUseStateForUnknown(),
```
Comment (henryrecker-pingidentity) [inline internal/resource/common/attributesources/attribute_sources_schema.go:25]: What if someone wants to update the data store ref?

File: internal/resource/config/oauth/accesstokenmapping/oauth_access_token_mappings_resource.go [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+	IssuanceCriteria             types.Object `tfsdk:"issuance_criteria"`
+}
+
+// GetSchema defines the schema for the resource.
+func (r *oauthAccessTokenMappingsResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	schema := schema.Schema{
+		Description: "Manages an OAuth Access Token Mapping",
+		Attributes: map[string]schema.Attribute{
+			"context": schema.SingleNestedAttribute{
+				Description: "The context of the OAuth Access Token Mapping. This property cannot be changed after the mapping is created.",
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/oauth/accesstokenmapping/oauth_access_token_mappings_resource.go:66]: Is it possible to enforce this in the provider?

File: internal/resource/config/oauth/accesstokenmapping/oauth_access_token_mappings_resource.go [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+	var err error
+
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	accessTokenMappingContext := &client.AccessTokenMappingContext{}
+	err = json.Unmarshal([]byte(internaljson.FromValue(plan.Context, true)), accessTokenMappingContext)
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/oauth/accesstokenmapping/oauth_access_token_mappings_resource.go:179]: What happens if plan.Context is null here?

File: internal/resource/config/oauth/accesstokenmapping/oauth_access_token_mappings_resource.go [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to create context object for request object:", err.Error())
+		return
+	}
+
+	accessTokenManagerRef, err := resourcelink.ClientStruct(plan.AccessTokenManagerRef)
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to create AccessTokenManagerRef for request object:", err.Error())
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/oauth/accesstokenmapping/oauth_access_token_mappings_resource.go:187]: Should there be a return here and in the below check? Or maybe a final HasError check after all these but before the API call?

File: internal/resource/config/oauth/accesstokenmapping/oauth_access_token_mappings_resource.go [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+	diags = readOauthAccessTokenMappingsResponse(ctx, apiReadOauthAccessTokenMappings, &state)
+	resp.Diagnostics.Append(diags...)
+
+	// Set refreshed state
+	diags = resp.State.Set(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+}
+
+// Update updates the resource and sets the updated Terraform state on success.
+func (r *oauthAccessTokenMappingsResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/oauth/accesstokenmapping/oauth_access_token_mappings_resource.go:248]: Same comments for this method as for the Create method

File: .gitignore [hunk 8]
```diff
@@ -37,6 +37,7 @@ terraform
 scripts/*
 scratch/*
 *__debug_bin*
+server-profiles/shared-profile/instance/bulk-config/data.json.subst
```
[the remark this answers is not in the document]
Comment (wesleymccollam) [reply .gitignore:40]: This file gets mounted when deploying the container locally. A remove command can't be added to the Makefile target due to the pipeline permissions.

File: internal/resource/config/oauth/accesstokenmapping/oauth_access_token_mappings_resource.go (hunk 5 above)
Comment (wesleymccollam) [reply internal/resource/config/oauth/accesstokenmapping/oauth_access_token_mappings_resource.go:179]: context is required, so it should always be in the plan

Comment (henryrecker-pingidentity) [reply internal/resource/config/oauth/accesstokenmapping/oauth_access_token_mappings_resource.go:179]: 👍 I misread and thought the modify plan said context_ref was not required.

File: internal/resource/common/resourcelink/resource_link_client_struct.go [hunk 9]
```diff
@@ -18,15 +18,10 @@ func ClientStruct(planObj basetypes.ObjectValue) (*client.ResourceLink, error) {
 	if !ok {
 		return nil, errors.New("object value missing \"id\" attribute when creating resource link client struct")
 	}
-	objLoc, ok := objValues["location"]
```
Comment (henryrecker-pingidentity) [inline internal/resource/common/resourcelink/resource_link_client_struct.go:21]: We still want to set location on the other resources, right?

File: internal/acctest/config/oauth/accesstokenmapping/oauth_access_token_mapping_test.go (hunk 2 above)
Comment (wesleymccollam) [reply internal/acctest/config/oauth/accesstokenmapping/oauth_access_token_mapping_test.go:86]: whoops! got lost in re-work

File: internal/resource/common/resourcelink/resource_link_client_struct.go (hunk 9 above)
Comment (wesleymccollam) [reply internal/resource/common/resourcelink/resource_link_client_struct.go:21]: will add boolean argument for this

**Review by henryrecker-pingidentity — APPROVED** [review]

**Pull request merged**